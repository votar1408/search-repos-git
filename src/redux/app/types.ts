export const LOADING: string = 'LOADING';

export interface AppLoadingAction {
    type: typeof LOADING;
    payload: boolean;
}

export type AppAction = AppLoadingAction;

export interface AppState {
    loading: boolean;
}
