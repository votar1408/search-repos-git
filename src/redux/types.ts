import {AppState} from './app/types';

export interface RootAppState {
    app: AppState;
}

interface RootState extends RootAppState {}

export const selectApp = (state: RootState) => state.app;
