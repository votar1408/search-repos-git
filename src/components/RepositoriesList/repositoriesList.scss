@import '../../assets/scss/variable.scss';

.repositories-container {
    max-width: 1000px;
    max-height: calc(100vh - 230px);
    overflow: hidden;
    overflow-y: auto;
    border: 1px solid $color-grey;
    border-radius: 5px;
    li:nth-child(even) {
        background-color: lighten($color: $color-main, $amount: 10);
        color: white;
    }
    li:last-child {
        border-bottom: none;
    }
}
