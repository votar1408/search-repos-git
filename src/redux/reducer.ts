import {combineReducers} from 'redux';

import {appReducer} from './app/reducer';

export const reducer = combineReducers({
    app: appReducer
});
