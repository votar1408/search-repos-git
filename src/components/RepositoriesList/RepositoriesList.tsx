import './repositoriesList.scss';

import React, {
    FC,
    ReactElement,
    useMemo
} from 'react';

import {useSelector} from 'react-redux';

import {texts} from '../../consts/texts';
import {RepositoryServerInfo} from '../../interfaces/ServerData';
import {selectApp} from '../../redux/types';
import {Repository} from './Repository';

export const RepositoriesList: FC = () => {
    const {repositories, page, loading} = useSelector(selectApp);
    const listForPage: RepositoryServerInfo[] = !repositories[page] ? repositories[page - 1] : repositories[page];

    const listItems: ReactElement | null = useMemo(() => {
        return listForPage?.length > 0 ? (
            <ul className="repositories-container">
                {listForPage.map((item: RepositoryServerInfo) => {
                    return <Repository key={item.id} item={item} />;
                })}
            </ul>
        ) : loading ? null : (
            <div className="empty-repositories">{texts.emptyList}</div>
        );
    }, [listForPage, loading]);

    return <>{listItems}</>;
};
