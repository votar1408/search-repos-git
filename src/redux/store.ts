/* eslint-disable no-underscore-dangle */
import {
    applyMiddleware,
    compose,
    createStore,
    Middleware
} from 'redux';
import thunk from 'redux-thunk';

import {reducer} from './reducer';

const middleware: [Middleware] = [thunk];

export const store = createStore(reducer, compose(applyMiddleware(...middleware)));
