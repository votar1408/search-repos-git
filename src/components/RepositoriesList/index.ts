import {RepositoriesList} from './RepositoriesList';

export {RepositoriesList};
