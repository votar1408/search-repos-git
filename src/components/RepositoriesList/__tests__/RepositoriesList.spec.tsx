import React from 'react';

import {
    mount,
    ReactWrapper,
    ShallowWrapper
} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import {Provider} from 'react-redux';
import {BrowserRouter as Router} from 'react-router-dom';
import createMockStore, {
    MockStoreCreator,
    MockStoreEnhanced
} from 'redux-mock-store';

import {RootAppState} from '../../../redux/types';
import {RepositoriesList} from '../RepositoriesList';

describe('RepositoriesList', () => {
    const initialState: RootAppState = {
        app: {
            loading: false,
            repositories: [
                [
                    {
                        id: 1,
                        name: 'test1',
                        description: 'some description 1',
                        owner: {
                            login: 'login1',
                            avatar_url: 'some url1'
                        }
                    },
                    {
                        id: 2,
                        name: 'test2',
                        description: 'some description 2',
                        owner: {
                            login: 'login2',
                            avatar_url: 'some url2'
                        }
                    }
                ]
            ],
            page: 1,
            searchValue: ''
        }
    };

    const mockStore: MockStoreCreator = createMockStore();

    let store: MockStoreEnhanced;
    let wrapper:
        | ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>
        | ShallowWrapper<{}, {}, React.Component<{}, {}, any>>;

    const getWrapper = (state: RootAppState) => {
        store = mockStore(state);
        return mount(
            <Router>
                <Provider store={store}>
                    <RepositoriesList />
                </Provider>
            </Router>
        );
    };

    beforeEach(() => {
        wrapper = getWrapper(initialState);
    });

    it('should render correctly', () => {
        expect(shallowToJson(wrapper as ShallowWrapper)).toMatchSnapshot();
    });

    it('renders without crashing', () => {
        expect(wrapper.find(RepositoriesList).length).toEqual(1);
    });

    it('should have 1 list element', () => {
        expect(wrapper.find('ul').length).toEqual(1);
    });

    it('should have 2 li element', () => {
        expect(wrapper.find('li').length).toEqual(2);
    });
});
