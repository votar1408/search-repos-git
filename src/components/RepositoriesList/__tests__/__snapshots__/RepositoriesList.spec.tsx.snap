// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoriesList should render correctly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <RepositoriesList>
        <ul
          className="repositories-container"
        >
          <Repository
            item={
              Object {
                "description": "some description 1",
                "id": 1,
                "name": "test1",
                "owner": Object {
                  "avatar_url": "some url1",
                  "login": "login1",
                },
              }
            }
            key="1"
          >
            <li
              className="repository"
            >
              <div
                className="repository__block"
                data-id="repository-block"
              >
                <span
                  className="repository__block_title"
                >
                  Name:
                </span>
                <span
                  className="repository__block_description"
                >
                  test1
                </span>
              </div>
              <div
                className="repository__block"
                data-id="repository-block"
              >
                <span
                  className="repository__block_title"
                >
                  Description:
                </span>
                <span
                  className="repository__block_description"
                >
                  some description 1
                </span>
              </div>
              <div
                className="repository__block"
                data-id="repository-block"
              >
                <span
                  className="repository__block_title"
                >
                  Owner:
                </span>
                <img
                  alt="logo"
                  className="repository__block_logo"
                  src="some url1"
                />
                <span
                  className="repository__block_description"
                >
                  login1
                </span>
              </div>
            </li>
          </Repository>
          <Repository
            item={
              Object {
                "description": "some description 2",
                "id": 2,
                "name": "test2",
                "owner": Object {
                  "avatar_url": "some url2",
                  "login": "login2",
                },
              }
            }
            key="2"
          >
            <li
              className="repository"
            >
              <div
                className="repository__block"
                data-id="repository-block"
              >
                <span
                  className="repository__block_title"
                >
                  Name:
                </span>
                <span
                  className="repository__block_description"
                >
                  test2
                </span>
              </div>
              <div
                className="repository__block"
                data-id="repository-block"
              >
                <span
                  className="repository__block_title"
                >
                  Description:
                </span>
                <span
                  className="repository__block_description"
                >
                  some description 2
                </span>
              </div>
              <div
                className="repository__block"
                data-id="repository-block"
              >
                <span
                  className="repository__block_title"
                >
                  Owner:
                </span>
                <img
                  alt="logo"
                  className="repository__block_logo"
                  src="some url2"
                />
                <span
                  className="repository__block_description"
                >
                  login2
                </span>
              </div>
            </li>
          </Repository>
        </ul>
      </RepositoriesList>
    </Provider>
  </Router>
</BrowserRouter>
`;
